### ISSUE TITLE
Component.glue_to can't assign component (only faces)

### ISSUE THREAD

**Issue by NeilBurkholder** [SketchUp/api-issue-tracker#111]
Please include the following:

1. SketchUp/LayOut Version: 2017 - 2018 Make & Pro
2. OS Platform: Windows 10

Describe briefly what you are doing and what is happening.
When placing a component in the UI is can be glued to either a face or another component.
When assigning the glued_to in the API however an error is displayed.

> Error: #<ArgumentError: wrong type - expected Sketchup::Face>

Describe what the expected result is.
Expected result is that the window component would be glued_to the wall compoent.
```ruby
window_component.glued_to = wall_component
```

Attach any relevant files.

# Feature Requests

1. Describe what you are trying to do on a high-level overview. What is your use-case?
2. If you have a spesific implementation in mind, please explain your reasoning to why it should be like that.

**Comment by thomthom** [comment]
This is a limitation of the current API. To support this we'd have to allow for an instance path to be provided in order to correctly attach to faces across nested instances.

**Comment by NeilBurkholder** [comment]
OK thanks. There are no work arounds as far as I know.

**Comment by DanRathbun** [comment]
+1 for implementing this. (It's basically in the GUI <-> API parity realm.)

**Comment by NeilBurkholder** [comment]
Just to clarify. I'm talking about component **instances**.

```ruby
window_component.glued_to == wall_component
true
window_component
#<Sketchup::ComponentInstance:0x00019f51426070>
wall_component
#<Sketchup::ComponentInstance:0x00019f51489d28>
window_component.glued_to = wall_component
ArgumentError: wrong type - expected Sketchup::Face
```

**Comment by Eneroth3** [comment]
> This is a limitation of the current API. To support this we'd have to allow for an instance path to be provided in order to correctly attach to faces across nested instances.

Judging from the model behavior and current getter return value no instance path is needed. It only glues to faces or instances in the same drawing context, not nested entities.

In time it could be cool to place a vase on a shelf in a bookcase, without be inside the bookcase component, and have it moved when the shelf moves, but just exposing what is already there to the Ruby API is enough in most cases.

**Comment by thomthom** [comment]
We want to add support for instance path to this like we did for attaching dimensions.

**Comment by Eneroth3** [comment]
Before reworking the core, would it be possible to just expose what is already supported to the API?

**Comment by thomthom** [comment]
Reworking the core? What do you mean? The core already uses instance paths - it's just the API that didn't expose that.

**Comment by Eneroth3** [comment]
Are you sure? From testing with the GUI it seems the core glues the instance to either a face or an instance, never a face inside of an instance. I don't think there is a path involved, just a reference to a single entity.

![2019-04-23_21h35_05](https://user-images.githubusercontent.com/5631725/56610482-c8bec980-660f-11e9-9c65-039d0d8de9f7.gif)

**Comment by Eneroth3** [comment]
Is this logged internally as something that could be implemented without having glue-to support instance paths? I need the existing functionality in SketchUp for a project but it isn't exposed to the API. I don't want to glue to any individual face inside of a group/component, just the group/component as a whole, as you already can do from the UI.

**Comment by thomthom** [comment]
hmm... maybe you are right. I might be confusing it with dimension-attachments.

**Comment by NeilBurkholder** [comment]
Yup I can verify that dimensions reference geometry inside a group/component but glued objects never reference entities inside the object they are glued to.

**Comment by thomthom** [comment]
I looked at the source, and it's not using instance paths at all. It's using a direct relationship to another entity. No 3d point references.

**Comment by Eneroth3** [comment]
This is great news! Then it should be relatively simple to expose it.

**Comment by thomthom** [comment]
And `Sketchup::Group` also needs these methods.

**Comment by Eneroth3** [comment]
This is also needed in the C API for exporters/importers, e.g. to map SketchUp gluing to 3ds Max parenting.

**Comment by NeilBurkholder** [comment]
Has this been looked at any more? I'm using the work around posted here.

https://forums.sketchup.com/t/bug-in-glued-to/74535/43?u=neil_burkholder

This work around is a real pain because the original component is replaced, and this causes issues with references to deleted entities.

**Comment by thomthom** [comment]
I'll bump it internally.

**Comment by NeilBurkholder** [comment]
A big thank you for implementing this in 2021.1
